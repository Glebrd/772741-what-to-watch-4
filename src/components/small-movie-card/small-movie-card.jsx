import React from "react";
import PropTypes from "prop-types";
import VideoPlayer from "../video-player/video-player.jsx";
import {MOVIE_WAIT} from "../../const.js";
import {ActionCreator} from "../../reducer/application/application";
import {Operation} from "../../reducer/data/data";
import {connect} from "react-redux";
import {ScreenType} from "../../const";
import {movieType} from "../../types";

class SmallMovieCard extends React.PureComponent {
  constructor(props) {
    super(props);

    this._timeout = null;
    this._videoRef = React.createRef();
  }
  _handleMovieCardHover() {
    this._timeout = setTimeout(() => {
      this._videoRef.current.play();
    }, MOVIE_WAIT);
  }

  _handleMovieCardUnhover() {
    clearTimeout(this._timeout);
    this._videoRef.current.load();
  }
  componentWillUnmount() {
    clearTimeout(this._timeout);
  }
  render() {
    const {smallMovieCard, onCardClick} = this.props;
    const {picture, title, videoPreview} = smallMovieCard;
    return (
      <article
        onMouseEnter={() => {
          this._handleMovieCardHover();
        }}
        onMouseLeave={() => {
          this._handleMovieCardUnhover();
        }}
        onClick={(evt) => {
          evt.preventDefault();
          onCardClick(smallMovieCard);
        }}
        className="small-movie-card catalog__movies-card">
        <div className="small-movie-card__image">
          <VideoPlayer
            videoRef={this._videoRef}
            poster={picture}
            videoPreview = {videoPreview}
          />
        </div>
        <h3 className="small-movie-card__title">
          <a className="small-movie-card__link" href="movie-page.html">{title}</a>
        </h3>
      </article>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  onCardClick(currentMovie) {
    dispatch(ActionCreator.setCurrentMovie(currentMovie));
    dispatch(ActionCreator.setCurrentScreen(ScreenType.MOVIE));
    dispatch(Operation.loadComments(currentMovie));
  },
});

SmallMovieCard.propTypes = {
  smallMovieCard: movieType,
  onCardClick: PropTypes.func,
};

export {SmallMovieCard};
export default connect(null, mapDispatchToProps)(SmallMovieCard);
