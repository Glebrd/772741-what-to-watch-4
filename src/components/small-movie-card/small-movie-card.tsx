import * as React from "react";
import VideoPlayer from "../video-player/video-player";
import {MOVIE_WAIT, AppRoute} from "../../const";
import {MovieType} from "../../types";
import {Link} from "react-router-dom";
import {RefObject} from "react";

interface Props {
  smallMovieCard: MovieType;
}
class SmallMovieCard extends React.PureComponent<Props> {
  private _videoRef: RefObject<HTMLVideoElement>;
  private _timeout: ReturnType<typeof setTimeout>;
  constructor(props) {
    super(props);

    this._timeout = null;
    this._videoRef = React.createRef();
  }
  componentWillUnmount() {
    console.log(this._timeout);
    clearTimeout(this._timeout);
  }
  _handleMovieCardHover() {
    this._timeout = setTimeout(() => {
      this._videoRef.current.play();
    }, MOVIE_WAIT);
  }

  _handleMovieCardUnhover() {
    clearTimeout(this._timeout);
    this._videoRef.current.load();
  }
  render() {
    const {smallMovieCard} = this.props;
    const {picture, title, videoPreview} = smallMovieCard;
    return (
      <article
        onMouseEnter={() => {
          this._handleMovieCardHover();
        }}
        onMouseLeave={() => {
          this._handleMovieCardUnhover();
        }}
        className="small-movie-card catalog__movies-card">
        <Link
          to={AppRoute.MOVIES + smallMovieCard.id}
        >
          <div className="small-movie-card__image">
            <VideoPlayer
              videoRef={this._videoRef}
              poster={picture}
              videoPreview={videoPreview}
            />
          </div>
        </Link>
        <h3 className="small-movie-card__title">
          <Link
            className="small-movie-card__link"
            to={AppRoute.MOVIES + smallMovieCard.id}
          >
            {title}
          </Link>
        </h3>
      </article>
    );
  }
}

export {SmallMovieCard};
export default SmallMovieCard;
