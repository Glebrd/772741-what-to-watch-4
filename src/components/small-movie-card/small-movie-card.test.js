import React from "react";
import renderer from "react-test-renderer";
import {SmallMovieCard} from "./small-movie-card.jsx";

const movie = {
  id: 29,
  title: `The Dark Knight`,
  picture: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
  genre: `Crime`,
  date: 2008,
  poster: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
  background: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
  rating: 8,
  scores: 1093,
  director: `Shaun Cathrall`,
  starring: [`Milissent Brahm`, `Gae Turton`, `Mitzi Baigrie`, `Tremayne Kilvington`, `Renado Dillon`],
  description: `When the menace known as the Joker wreaks havoc and chaos on the people of Gotham, Batman must accept one of the greatest psychological and physical tests of his ability to fight injustice.`,
  videoPreview: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
};

const onCardClick = () => {};
const onMovieCardHover = () => {};

it(`Catalog renders correctly`, () => {
  const tree = renderer
    .create(
        <SmallMovieCard
          smallMovieCard={movie}
          onCardClick={onCardClick}
          onMovieCardHover={onMovieCardHover}
        />
        , {
          createNodeMock: () => {
            return {};
          }
        }).toJSON();

  expect(tree).toMatchSnapshot();
});
