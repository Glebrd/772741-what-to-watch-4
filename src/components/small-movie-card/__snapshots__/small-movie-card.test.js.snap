// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Catalog renders correctly 1`] = `
<article
  className="small-movie-card catalog__movies-card"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <a
    href="/films/29"
    onClick={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <video
        height="175"
        muted={true}
        poster="img/fantastic-beasts-the-crimes-of-grindelwald.jpg"
        src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
        width="280"
      />
    </div>
  </a>
  <h3
    className="small-movie-card__title"
  >
    <a
      className="small-movie-card__link"
      href="/films/29"
      onClick={[Function]}
    >
      The Dark Knight
              
    </a>
  </h3>
</article>
`;
